cep-20.2.md

```diff
@@ -0,0 +1,651 @@
+
+```
+test: [TestElement]
+```
+
+where the different test elements are defined as follows:
+
+#### Command test element
+
+```yaml
+# script to execute
+script: string | [string]
+# optional extra requirements
+extra_requirements:
+  # extra requirements with build_platform architecture (emulators, ...)
+  build: [MatchSpec]
+  # extra run dependencies
+  run: [MatchSpec]
+# extra files to add to the package for the test
+# Alternatively, use ${{ RECIPE_DIR }} or ${{ SOURCE_DIR }} and a single list?
+# or $RECIPE_DIR / $SRC_DIR ...
+files:
+  # files from $SRC_DIR
+  source: [glob]
+  # files from $RECIPE_DIR
+  recipe: [glob]
+```
+
+#### Import test element
+
+```yaml
+# list of imports to try
+imports: [string]
+```
+
+#### Downstream test element
+
+```yaml
+downstream: MatchSpec
+```
+
+## Outputs section
+
+conda-build has very hard to understand behavior with multiple outputs. We propose some drastic simplifications.
+Each output in the new format has the same keys as a "top-level" recipe.
+
+Values from the top-level `build` and `about` section are merged into the outputs. The `version` from the top-level
+`package` is also merged into each output. The top-level name is ignored.
+
+```yaml
+outputs:
+  - package:
+      name: string
+      version: string (defaults to top-level version)
+    build:
+      # this package can only be used by other _outputs_ as build cache (for split packages)
+      # in that way it replaces the top-level build
+      # alternative names: package: true/false, output_package: true/false
+      cache_only: bool (defaults to false)
+      # cache from refers to the output from which the cache should be expanded into the WORK DIR
+      cache_from: string | [string]
+      # everything the same as top level
+      script: ...
+    requirements:
+      # same definition as top level
+
+    # as top level, by default merged from outer recipe
+    about:
+    source:
+    test:
+```
+
+A cache only package does only package the `$WORK_DIR` (similar to how top-level packages work today) and each output that depends on a cache package via `cache_from`
+will get a prepared `$WORK_DIR` with the build artifacts from the previously executed package.
+
+> **Note**
+> Should run-dependencies from the cache_only package be inherited? That would make `run_exports` work better ...
+
+## About section
+
+```yaml
+about:
+  # The license
+  # We can enforce SPDX for license?
+  license: string
+  # license files
+  license_file: path | [path] (relative paths are found in source directory _or_ recipe directory)
+  # a summary of what this package does
+  summary: string
+  # a longer description of what this package does (should we allow referencing files here?)
+  description: string
+  # URL to the homepage
+  home: url
+  # url to the repository - rename?
+  dev_url: url
+  # URL to the docs
+  doc_url: url
+  # unsure if we should keep this one
+  license_url: url
+
+  # REMOVED:
+  # prelink_message:
```

Comment:
this bit here was added for EULA like stuffs that needs confirmation from the user, I believe NVIDIA or/and intel folks use it? or at least they were keen to have it. I would be careful in removing this one :)

Reply:
Yes I 100% agree. We have to keep this.

Reply:
I suppose we could use prelink scripts? @jakirkham can comment for the nvidia stuff.

Reply:
They aren't using it on conda-forge (I checked)

Reply:
I am also keen on dropping the `post-link`, `pre-link`, and `post-unlink` keys in `build`.

Reply:
The keys or the scripts themselves or both? IDK actually what those keys do in a recipe.

Reply:
The keys in the recipe. We would keep the files.

Reply:
Prelink_message is mainly for eula purposes that the user needs to accept it, it is not related to the prelink and postlink script that are, as far I'm concerned, deprecated

Reply:
> prelink and postlink script that are, as far I'm concerned, deprecated

post-link and pre-unlink should only be deprecated if similar functionality is offered elsewhere.

> The keys in the recipe. We would keep the files.

Personally, I'm not a big fan of the "some file relative to `meta.yaml`/`recipe.yaml` triggers some implicit behavior" thing.
(I'd even go as far as removing the implicit invocation of `build.sh`/`bld.bat` -- but that might upset people :shrug:.)

Reply:
Was there a resolution to this for packages that want / need to display some kind of EULA message or even want to require some kind of interaction from a user, i.e. accepting a EULA?

NVIDIA legal seems to be appeased currently, but from experience these decisions can change, and there will likely be uses outside of NVIDIA packages as well.

Reply:
It feels like the "if a file named in this particular way is found then we do X" behaviour will stay, but the keys to change how that file is named will not... If we do so, I'd rather we do it like we treat activation scripts (their _target_ location is the trigger, not the source one); but that doesn't work well for pre-link & post-unlink because the files will be gone.